core-tools.markdown

```diff
@@ -0,0 +1,345 @@
+the types of most expressions, it is recommended that top-level definitions
+are annotated with their types.
+
+We can remedy that by adding the following line above our main definition:
+
+```hs
+main :: IO ()
+```
+
+Now our `hello.hs` source file looks like this:
+
+```hs
+main :: IO ()
+main = putStrLn "Hello, world!"
+```
+
+And now (after deleting the build artifact files) GHC will compile `hello.hs`
+without warnings.
+
+`-Wall` emits many useful warnings that can easily be bugs,
+including warnings about name shadowing, unused variables, and more.
+
+Note that `-Wall` does not stop GHC from compiling your program.
+If this is something you'd like to do, specifying the flag `-Werror`
+will convert warnings to errors and will halt compilation when a warning is emitted.
+
+#### Optimisations
+
+Another very useful flag is `-O`. This flag asks GHC to compile the program with optimisations and code improvements at the cost of taking longer to compile.
+This will often make the program run much faster and perform more aggressive optimisations such as inlining and specialisation.
+For even more optimisations that may take significantly longer, `-O2` is also available.
+
+## An interactive environment
+
+GHC provides an interactive environment in a form of a
+Read-Evaluate-Print Loop (REPL) called GHCi.
+To enter the environment run the program `ghci`.
+
+```sh
+➜ ghci
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+ghci> 
+```
+
+It provides an interactive prompt where Haskell expressions can be written and
+evaluated.
+
+For example:
+
+```sh
+ghci> 1 + 1
+2
+ghci> putStrLn "Hello, world!"
+Hello, world!
+```
+
+We can define new names:
+
+```sh
+ghci> double x = x + x
+ghci> double 2
+4
+```
+
+We can import Haskell source files using the `:load` command:
+
+```sh
+ghci> :load hello.hs
+[1 of 1] Compiling Main             ( hello.hs, interpreted )
+Ok, one module loaded.
+ghci> main
+Hello, Haskell!
+```
+
+As well as import library modules:
+
+```sh
+ghci> import Data.Bits
+ghci> shiftL 32 1
+64
+ghci> clearBit 33 0
+32
+```
+
+We can even ask what the type of an expression is using the `:type` command:
+
+```sh
+λ> :type putStrLn
+putStrLn :: String -> IO ()
+```
+
+To exit `ghci`, use the `:quit` command (or `:q` for short)
+
+```sh
+ghci> :quit
+Leaving GHCi.
+```
+
+A more thorough introduction to GHCi can be found in the [GHC user guide](https://downloads.haskell.org/~ghc/9.0.1/docs/html/users_guide/ghci.html).
+
+### Using external packages in ghci
+
+By default, ghci can only load and use packages that are [included with the GHC installation](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/9.0.1-notes.html#included-libraries).
+
+However, users of [cabal-install](https://www.haskell.org/cabal) and [stack](http://haskellstack.org) can download and load external packages very easily using the following commands:
+
+Cabal-install:
+
+```sh
+cabal repl --build-depends random
```

Comment:
I suggest adding or modifying the example to show how to load more than one package because the syntax for that is not immediately obvious from that for one package. (Both cabal and stack.)